partiql-ast/src/experimental/ast.rs

```diff
@@ -11,16 +11,67 @@
 
 // TODO Add documentation.
 
+use partiql_source_map::location::BytePosition;
 use rust_decimal::Decimal as RustDecimal;
+use std::collections::HashMap;
 use std::fmt;
 
+/// Provides the required methods for AstNode conversations.
+pub trait ToAstNode {
+    /// Wraps the `self` to an [AstNode] and returns an `AstNodeBuilder` for
+    /// further [AstNode] construction.
+    /// ## Example:
+    /// ```
+    /// use partiql_ast::experimental::ast::{Span, SymbolPrimitive, ToAstNode};
+    /// use partiql_source_map::location::BytePosition;
+    ///
+    /// let p = SymbolPrimitive {
+    ///     value: "symbol2".to_string()
+    ///  };
+    ///
+    /// let node = p
+    ///     .to_node()
+    ///     .span(Span {
+    ///         begin: BytePosition::from(12),
+    ///             end: BytePosition::from(1),
+    ///     })
+    ///     .build()
+    ///     .expect("Could not retrieve ast node");
+    /// ```
+    fn to_node(self) -> AstNodeBuilder<Self>
+    where
+        Self: Clone,
+    {
+        AstNodeBuilder::default().node(self).clone()
+    }
+}
+
+/// Implements [ToAstNode] for all types within this crate, read further [here][1].
+/// [1]: https://doc.rust-lang.org/book/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods
+impl<T> ToAstNode for T {}
+
+/// Represents an AST node. [AstNode] uses [derive_builder][1] to expose a Builder
+/// for creating the node. See [ToAstNode] for more details on the usage.
+/// [1]: https://crates.io/crates/derive_builder
+#[derive(Builder, Clone, Eq, PartialEq, Debug)]
+pub struct AstNode<T> {
+    pub node: T,
+    #[builder(setter(strip_option), default)]
+    pub span: Option<Span>,
+}
+
+/// Represents the beginning and the end of a `Span` in the source code
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Span {
+    pub begin: BytePosition,
+    pub end: BytePosition,
+}
+
```

Comment:
Will consider adding type aliases as recommended [here](https://github.com/partiql/partiql-lang-rust/issues/81) post reviewing this PR.